keras/layers/variational.py

```diff
@@ -0,0 +1,125 @@
+    regularizer_scale: By default the regularization is already properly
+        scaled if you use binary or categorical crossentropy cost functions.
+        In most cases this regularizers should be kept fixed at one.
+
+    Notes:
+    ------
+    During training we output ONE sample from the reparameterized
+    distribution given by "mean + std * noise".
+    For testing, the output of this layer is deterministic given by "mean".
+
+    References:
+    -----------
+    - [Auto-Encoding Variational Bayes](http://arxiv.org/pdf/1312.6114v10.pdf)
+
+    """
+    def __init__(self, output_dim, init='glorot_uniform',
+                 activation='tanh',
+                 weights=None, input_dim=None, regularizer_scale=1,
+                 prior_mean=0, prior_logsigma=1, **kwargs):
+        self.prior_mean = prior_mean
+        self.prior_logsigma = prior_logsigma
+        self.regularizer_scale = regularizer_scale
+        self.init = initializations.get(init)
+        self.activation = activations.get(activation)
+        self.output_dim = output_dim
+        self.initial_weights = weights
+        self.input_dim = input_dim
+        if self.input_dim:
+            kwargs['input_shape'] = (self.input_dim,)
+        self.input = K.placeholder(ndim=2)
+        super(VariationalDense, self).__init__(**kwargs)
+
+    def build(self):
+        input_dim = self.input_shape[-1]
+
+        self.W_mean = self.init((input_dim, self.output_dim),
+                                name='{}_W_mean'.format(self.name))
+        self.b_mean = K.zeros((self.output_dim,), name='{}_b_mean'.format(self.name))
+        self.W_logsigma = self.init((input_dim, self.output_dim),
+                                    name='{}_W_logsigma'.format(self.name))
+        self.b_logsigma = K.zeros((self.output_dim,), name='{}_b_logsigma'.format(self.name))
+
+        self.trainable_weights = [self.W_mean, self.b_mean, self.W_logsigma,
+                                  self.b_logsigma]
+
+        self.regularizers = []
+        reg = self.get_variational_regularization(self.get_input())
+        self.regularizers.append(reg)
+
+    def get_variational_regularization(self, X):
+        mean = self.activation(K.dot(X, self.W_mean) + self.b_mean)
+        logsigma = self.activation(K.dot(X, self.W_logsigma) + self.b_logsigma)
+        return GaussianKL(mean, logsigma,
+                          regularizer_scale=self.regularizer_scale,
+                          prior_mean=self.prior_mean,
+                          prior_logsigma=self.prior_logsigma)
+
+    def get_mean_logsigma(self, X):
+        mean = self.activation(K.dot(X, self.W_mean) + self.b_mean)
+        logsigma = self.activation(K.dot(X, self.W_logsigma) + self.b_logsigma)
+        return mean, logsigma
+
+    def _get_output(self, X, train=False):
+        mean, logsigma = self.get_mean_logsigma(X)
+        if train:
+            # infer batch size for theano backend
+            if K._BACKEND == 'theano':
```

Comment:
A minor comment: it is suspicious that you have to put "if K._BACKEND == 'theano':" right in the application code as opposed to somehow solving this on a backend side (like patching random_normal). But I haven't looked whats the problem exactly, so maybe it is legitimate here.